### PR TITLE
persist: add initial simple invariant-based random tests

### PR DESCRIPTION
**Pull request by danhhz**
This module generates random traffic against the external API of the
persist crate. It records the requests sent and responses received and
afterward checks this history against the API invariants.

API traffic is modeled as a series of [Input]s. These are created by a
[Generator], which is totally deterministic given a seed. The [Input]s
are handed to an implementation of [Runtime] which executes the embedded
[Req] and saves the [Res] wrapped up with metadata about the execution
into a [Step]. These steps collectively form a history, which is handed
to [Validator] to check for API invariant violations (representing a bug
in the persist crate).

The initial Runtime implementation is [Direct].

Several of these components can be configured. Particularly interesting
combinations of these are committed as unit tests. By default, these
unit tests all select and use a random seed for traffic generation. This
runs the test randomly once and is of limited usefulness.

    cargo test -p persist -- nemesis

The seed is printed when the test is run and override-able, so if a test
does fail (persist violated one of its invariants aka a bug), the
previous traffic can be exactly replayed. For a deterministic runtime
(the "direct_*" variants) this results in exactly the same execution
sequence and can be used to verify the bug fix.

    NEMESIS_SEED=16385522461931935384 cargo test -p persist -- failing_test_name

If the runtime is not deterministic, the seed can still be fixed to
increase the likelihood of ticking the same bug again. The cargo-stress
crate, which runs a unit test (or set of tests) in a tight loop until
failure, is useful for this.

    NEMESIS_SEED=16385522461931935384 cargo stress -p persist -- failing_test_name

Finally, the cargo-stress crate is also useful without fixing the seed
to search for new invariant violations. This repeatedly runs all nemesis
tests, each with new random traffic to maximize the chances of finding
an interesting history.

    cargo stress -p persist -- nemesis

### CODE REVIEW

**Comment by danhhz** [comment]
@philip-stoev and @mjibson, adding you as optional reviewers because this seems like it's in your wheelhouse and I'd love your thoughts. This is currently just a friday project, so no hurry.

I still owe some comments in the code, but I wanted to get this out today for an initial look and I ran out of time. It's already found 3 new bugs (and one we knew about)!

File: src/persist/src/nemesis/generator.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    pub fn gen(&mut self) -> Input {
+        let req_id = ReqId(self.req_id);
+        self.req_id += 1;
+        let mut rng = SmallRng::seed_from_u64(self.seed + req_id.0);
+        let req = loop {
+            break match rng.gen_range(0..=3) {
+                0 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let key = rng.gen_range('a'..'e').to_string();
+                    let ts = rng.gen_range(0..100);
```
Comment (philip-stoev) [inline src/persist/src/nemesis/generator.rs:43]: This will distribute timestamps uniformly over the space, but I thought timestamps are monotonically increasing? Can you get an event for a previous timestamp after a more recent one has been seen?

File: src/persist/src/nemesis/generator.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                1 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let ts = rng.gen_range(0..100);
+                    Req::Seal(SealReq { stream, ts })
+                }
+                2 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let snap = SnapshotId(self.snap_id);
+                    self.snap_id += 1;
+                    Req::TakeSnapshot(TakeSnapshotReq { stream, snap })
```
Comment (philip-stoev) [inline src/persist/src/nemesis/generator.rs:59]: Are you allowed to have any other operations over snapshots besides Taking and then restoring one (e.g. delete it, compact it, whatever)? The test does not recycle snapshot ids, so I assume Mz in production will also not be recycling snapshot ids as it is running?

File: src/persist/src/nemesis/generator.rs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        let req_id = ReqId(self.req_id);
+        self.req_id += 1;
+        let mut rng = SmallRng::seed_from_u64(self.seed + req_id.0);
+        let req = loop {
+            break match rng.gen_range(0..=3) {
+                0 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let key = rng.gen_range('a'..'e').to_string();
+                    let ts = rng.gen_range(0..100);
+                    let diff = if rng.gen_bool(0.9) { 1 } else { -1 };
```
Comment (philip-stoev) [inline src/persist/src/nemesis/generator.rs:44]: Does any part of the persistence code care about the contents of "key" and "diff (e.g. is there compaction, deduplication, etc?) If yes, then a non-uniform distribution over the keys may be useful, as well as diffs outside of -1 and 1 .

File: src/persist/src/nemesis/generator.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    pub fn gen(&mut self) -> Input {
+        let req_id = ReqId(self.req_id);
+        self.req_id += 1;
+        let mut rng = SmallRng::seed_from_u64(self.seed + req_id.0);
+        let req = loop {
+            break match rng.gen_range(0..=3) {
+                0 => {
+                    let stream = rng.gen_range(0..5).to_string();
+                    let key = rng.gen_range('a'..'e').to_string();
```
Comment (philip-stoev) [inline src/persist/src/nemesis/generator.rs:42]: Is this thing going to be fired against actual S3 or a S3-compatible local service? If so, some larger keys may benefit so that the PUT and GET operations start having more varied execution times in order to shake out any race conditions.

**Review by philip-stoev — APPROVED** [review]
Thank you for adding me as a reviewer, reading through this was very instructional.

All looks good -- my biggest suggestion would be to make the distribution a first-class citizen so that one can plug a non-uniform/zipfian distribution and then have the entire thing naturally shift to to be hammering the same key over and over again, with many write operations per snapshot taken/restored.

File: src/persist/src/nemesis/validator.rs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                    let mut expected: Vec<((String, ()), u64, isize)> = self
+                        .writes_by_seqno
+                        .range((stream.to_owned(), SeqNo(0))..(stream.to_owned(), SeqNo(res.seqno)))
+                        .map(|(_, v)| v)
+                        .cloned()
+                        .collect();
+                    expected.sort();
+                    // TODO: We need to consolidate and account for since here once
+                    // that's supported in persist.
+                    if actual != expected {
```
Comment (philip-stoev) [inline src/persist/src/nemesis/validator.rs:140]: Once straight equality of actual v.s. expected becomes untenable, can you direct the actual and the expected to two new empty materialized arrangements and then read from them to obtain two comparable resultsets?

**Review by philip-stoev — APPROVED** [review]

File: src/persist/src/nemesis/generator.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            snap_id: 0,
+        }
+    }
+
+    pub fn gen(&mut self) -> Input {
+        let req_id = ReqId(self.req_id);
+        self.req_id += 1;
+        let mut rng = SmallRng::seed_from_u64(self.seed + req_id.0);
+        let req = loop {
+            break match rng.gen_range(0..=3) {
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/persist/src/nemesis/generator.rs:39]: Fully agree about this. My first randomizer did everything with equal probability and it was ok but could have been better if it used a different distribution. For my second one, I allowed each operation to indicate how often it'd like to be performed. See https://github.com/cockroachdb/cockroach/blob/4b63eb40096723ad025b71cee8edbb0db8c1975b/pkg/internal/sqlsmith/scalar.go#L21 for example. This was important to generate more realistic workloads.

File: src/persist/src/nemesis/generator.rs (hunk 3 above)
Comment (danhhz) [reply src/persist/src/nemesis/generator.rs:52]: Yup! Added non-uniform distribution of keys to the big TODO.

Genuine question, do we need diffs other than -1 and 1? AFAIK that's the overwhelming majority of usage and I'd been thinking that'd be enough to stress compaction/consolidation as well as things cancelling out.

File: src/persist/src/nemesis/generator.rs (hunk 2 above)
Comment (danhhz) [reply src/persist/src/nemesis/generator.rs:67]: The snapshot IDs are actually just an artifact of the structure of the nemesis test (and one of the things I meant to call out when I mentioned "I still owe some comments in the code" in my request for your review). A snapshot user just gets an in-memory object which it uses directly, but I wanted to stress race conditions between snapshots and varies write/compaction/etc operators as well as eventually the edge cases around the still TODO disk-backed LRU cache for the persisted data. The snapshot IDs are what I came up with to do this while keeping Generator deterministic.

File: src/persist/src/nemesis/validator.rs (hunk 5 above)
Comment (danhhz) [reply src/persist/src/nemesis/validator.rs:140]: Hmmm great question! Historically, I've tried to stress edge cases by writing test variants with different generator configurations that target the interactions I'm worried about instead of running things for a long time. You're right that, in this case, internal block sizes etc are going to make it more necessary to stress large data sizes. Perhaps if we make all those tunable, we can keep data sizes small enough that the Validator can deal with an entire history in memory?

If not, you're correct that an arrangement would be a great way to check this. However, there's actually a possibility that we'll end up using some of the differential dataflow trace stuff internally, which could mean we might lose the opportunity to catch a bug that affected both the real implementation and the verifier. A hash is also an interesting idea but I don't immediately see how to make the negations cancel out.

My current guess is that we'll push on this as much as possible by tweaking internal parameters around byte sizes and leave large data sizes for a separate test designed specifically for that, but I'm open to other opinions. Thoughts?

**Comment by ruchirK** [comment]
Going to review this tomorrow first thing!

File: src/persist/src/nemesis/generator.rs (hunk 3 above)
Comment (mjibson) [reply src/persist/src/nemesis/generator.rs:52]: Might consider adding 2 and -2 because there's some "subtract by one" code and so this might exercise some more edge cases.

File: src/persist/src/nemesis/generator.rs (hunk 1 above)
Comment (philip-stoev) [reply src/persist/src/nemesis/generator.rs:51]: >  A user with some sort of dataflow input handle gets a "capability" to introduce data at any timestamp they like. As they go along, they "downgrade" the capability to a higher timestamp, which is a promise that they're done introducing data below that timestamp.

Thanks for the explanation, so before you retire timestamp X, you are allowed to introduce events for X , X - 1 and all the way down to timestamp 0. Good to know, I thought timely will straight up panic on you if you introduced an event for X - 1 or lower. Will keep that in mind when it comes to testing future stuff.

File: src/persist/src/nemesis/validator.rs (hunk 5 above)
Comment (philip-stoev) [reply src/persist/src/nemesis/validator.rs:140]: Understood. When I said "untenable", I did not mean running out of memory, I mostly meant situations where, due to consolidation or another future operation/optimization, the "actual" ends up with a -2 diff while the "expected" had two -1 diffs or the timestamps get rewritten, that sort of suff.

File: src/persist/src/nemesis/mod.rs [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+//! search for new invariant violations. This repeatedly runs all nemesis tests,
+//! each with new random traffic to maximize the chances of finding an
+//! interesting history.
+//!
+//! ```shell
+//! cargo stress -p persist -- nemesis
+//! ```
+
+// TODO
+// - Variants with FileBuffer and FileBlob
```
Comment (ruchirK) [inline src/persist/src/nemesis/mod.rs]: we'll also want to add support for destroying streams once we build that out

File: src/persist/src/nemesis/generator.rs (hunk 3 above)
Comment (ruchirK) [reply src/persist/src/nemesis/generator.rs:52]: I would advocate for supporting diff = 0 in write sync, and a mode where we sample diffs between [+1000, -1000] or something like that. I agree that +1 / -1 diffs are expected usage for tables and sources but I think its different when persisting arbitrary operators

File: src/persist/src/nemesis/generator.rs (hunk 1 above)
Comment (danhhz) [reply src/persist/src/nemesis/generator.rs:51]: Right. Before you've retired anything, you can introduce data at [0, inf). Then if you retire X, you're allowed to introduce data at [X, inf). I'm actually not sure what timely does here if you use it wrong, it very well might panic. However, in persist, we already have to handle errors resulting from io failures, etc, so we may as well more gracefully handle incorrect API usage and return an error.

**Review by ruchirK — APPROVED** [review]
LGTM! I added some small nits but I read through the whole thing carefully and pretty much just agreed with everything :D

File: src/persist/src/nemesis/validator.rs (hunk 5 above)
Comment (danhhz) [reply src/persist/src/nemesis/validator.rs:140]: Oh! Yeah, for that case, I'd rather reimplement the consolidation logic in an obvious (and possibly slower) way inside validate and then call it on actual and expected before comparing them. It's nice for validate's code to be independent of what it's testing

File: src/persist/src/indexed/mod.rs [hunk 8]
```diff
@@ -458,6 +461,15 @@ pub struct IndexedSnapshot<K, V>(
     TraceSnapshot<K, V>,
 );
 
+impl<K, V> IndexedSnapshot<K, V> {
+    /// Returns the SeqNo at which this snapshot was run.
+    ///
+    /// All writes assigned a seqno < this are included.
+    pub fn seqno(&self) -> SeqNo {
+        self.0 .0
```
[the remark this answers is not in the document]
Comment (danhhz) [reply src/persist/src/indexed/mod.rs:469]: Dunno! rustfmt is doing this

File: src/persist/src/nemesis/generator.rs (hunk 1 above)
Comment (ruchirK) [reply src/persist/src/nemesis/generator.rs:51]: timely panics if you try to downgrade capabilities incorrectly

File: src/persist/src/nemesis/mod.rs (hunk 7 above)
Comment (danhhz) [reply src/persist/src/nemesis/mod.rs]: Added

File: src/persist/src/nemesis/generator.rs (hunk 3 above)
Comment (danhhz) [reply src/persist/src/nemesis/generator.rs]: Okay yeah, this is great! Changed this to a uniform distribution in [-2,2]. Operator persistence is a great point, added a TODO to think about new edge cases to test for that.

**Review by danhhz — COMMENTED** [review]
Thanks for the reviews everyone!

**Comment by danhhz** [comment]
Test failure looks like a flake, retrying them.

```
error: Failed to clean up state at shut down: deleting S3 buckets: 1 errorsdeleting S3 buckets: 1 errors: Request ID: None Body: <?xml version="1.0" encoding="UTF-8"?>
--
  | <Error><Code>InternalError</Code><Message>We encountered an internal error. Please try again.</Message><RequestId>K6R5CF4HV83GAK9C</RequestId><HostId>IEx6c73O6LhGjyLIQH1gGjw2tOlsOr2q7VzWXfocVNAcdp3u2SBVUYh9/RQ2XK7fGivdUyD059M=</HostId></Error>
  |  
  | 1 errors were encountered during execution
  | files involved: esoteric/s3.td
  | Failed! giving up: testdrive-svc --aws-region=us-east-2 --ci-output '*.td' 'esoteric/*.td'
```
